### PR TITLE
Feature/molvis sequence

### PR DESCRIPTION
**Pull request by martinfalk**

### CODE REVIEW

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h [hunk 1]
```diff
@@ -151,6 +173,29 @@ namespace molvis {
  */
 class IVW_MODULE_MOLVISBASE_API MolecularStructure {
 public:
+    /**
+     * \brief data structure holding the atom indices of a single backbone segment including
+     * dihedral angles
+     */
+    struct BackboneSegment {
+        bool valid() const {
+            return ca.has_value() && n.has_value() && c.has_value() && o.has_value();
```
Comment (petersteneteg) [inline molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:182]: Why are they optional if they have to be there to be valid? Are they only optional during construction then?

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h [hunk 2]
```diff
@@ -151,6 +173,29 @@ namespace molvis {
...
+        size_t resId;
+        size_t chainId;
+
+        std::optional<size_t> ca;
+        std::optional<size_t> n;
+        std::optional<size_t> c;
+        std::optional<size_t> o;
+
+        // dihedral angles
+        double phi = 0.0;
```
Comment (petersteneteg) [inline molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:194]: Should these not be calculated then? What invariants you you have

File: molvis/molvisbase/src/datastructures/molecularstructure.cpp [hunk 3]
```diff
@@ -191,6 +213,140 @@ void MolecularStructure::verifyData() const {
...
+                        if (!seg.o.has_value()) seg.o = atomIdx;
+                        break;
+                    default:
+                        break;
+                }
+                if (seg.valid()) break;
+            }
+
+            if (!seg.valid()) {
+                ++invalidCount;
```
Comment (petersteneteg) [inline molvis/molvisbase/src/datastructures/molecularstructure.cpp:283]: What is invalid count used for?

File: molvis/molvisbase/src/datastructures/molecularstructure.cpp [hunk 4]
```diff
@@ -191,6 +213,140 @@ void MolecularStructure::verifyData() const {
...
+            }
+            segments.push_back(seg);
```
Comment (petersteneteg) [inline molvis/molvisbase/src/datastructures/molecularstructure.cpp:285]: Why add invalid segments

Comment (petersteneteg) [reply molvis/molvisbase/src/datastructures/molecularstructure.cpp:285]: why not compute the angles here?

File: molvis/molvisbase/src/datastructures/molecularstructure.cpp [hunk 5]
```diff
@@ -191,6 +213,140 @@ void MolecularStructure::verifyData() const {
...
+        }
+        chainSegments_.emplace(chain.first, std::move(segments));
+    }
+
+    computeDihedralAngles();
```
Comment (petersteneteg) [inline molvis/molvisbase/src/datastructures/molecularstructure.cpp:290]: Avoid having functions like these that only works with side effect unless absolutely needed. This is just a renamed initialise function.

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h (hunk 1 above)
Comment (martinfalk) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:182]: rename to `complete()`, add also add `empty()`

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h (hunk 2 above)
Comment (martinfalk) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:194]: maybe angles optional

File: molvis/molvisbase/src/datastructures/molecularstructure.cpp (hunk 4 above)
Comment (martinfalk) [reply molvis/molvisbase/src/datastructures/molecularstructure.cpp:285]: invalid segments are still part of the background and are therefore needed. Might be that the underlying data source was incomplete...

Comment (martinfalk) [reply molvis/molvisbase/src/datastructures/molecularstructure.cpp:285]: The computations cannot be done here as it requires the previous and next segment.

File: molvis/molvisbase/src/datastructures/molecularstructure.cpp (hunk 3 above)
Comment (martinfalk) [reply molvis/molvisbase/src/datastructures/molecularstructure.cpp]: removed counter

File: molvis/molvisbase/src/util/molvisutils.cpp [hunk 6]
```diff
@@ -232,6 +256,54 @@ std::shared_ptr<Mesh> createMesh(const MolecularStructure& s) {
...
+        } else {
+            tb(H("Residue"), fmt::format("{}", atoms.residueIds[atomIndex]));
+        }
+    }
+    if (!atoms.chainIds.empty()) {
+        const auto chainId = atoms.chainIds[atomIndex];
+        if (auto chain = findChain(s.data(), chainId)) {
+            tb(H("Chain"), fmt::format("{} (id: {})", s.chains()[chainId].name, chainId));
+        } else {
+            tb(H("Chain"), fmt::format("{}", chainId));
```
Comment (petersteneteg) [inline molvis/molvisbase/src/util/molvisutils.cpp:295]: think the table building will convert to str automatically

**Review by petersteneteg — APPROVED** [review]

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h [hunk 7]
```diff
@@ -204,6 +258,14 @@ class IVW_MODULE_MOLVISBASE_API MolecularStructure {
      */
     const std::vector<size_t>& getChainResidues(size_t chainId) const;
 
+    /**
+     * query a chain matching \p chainId for its backbone segments.
```
[the remark this answers is not in the document]
Comment (martinfalk) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:262]: marks the next word as a function argument/parameter, might even add a link to the corresponding `@param`

File: molvis/molvisbase/include/inviwo/molvisbase/util/molvisutils.h [hunk 8]
```diff
@@ -95,9 +103,27 @@ getAtomicNumbers(const std::vector<std::string>& fullNames);
...
+                                                           uint32_t startId = 0);
+
+/**
+ * \brief create a tool tip for the given \p atom of molecular structure \p s
+ *
+ * @param s      molecular structure
+ * @param atom   index of the atom
+ * @return tool tip with information on atomic element, residue, chain, etc.
+ */
+std::string IVW_MODULE_MOLVISBASE_API createToolTip(const MolecularStructure& s, int atomIndex);
```
[the remark this answers is not in the document]
Comment (martinfalk) [reply molvis/molvisbase/include/inviwo/molvisbase/util/molvisutils.h:126]: changed. Was inspired by tool tip function in DataFrame utils.

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h [hunk 9]
```diff
@@ -151,6 +175,36 @@ namespace molvis {
...
+     */
+    struct BackboneSegment {
+        bool complete() const {
+            return ca.has_value() && n.has_value() && c.has_value() && o.has_value();
+        }
+        bool empty() const {
+            return !ca.has_value() && !n.has_value() && !c.has_value() && !o.has_value();
+        }
+
+        size_t resId;
```
[the remark this answers is not in the document]
Comment (martinfalk) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:194]: Good suggestion. Will not be part of this pull request though.

[the remark this answers is not in the document]
Comment (petersteneteg) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:194]: sure

File: molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (petersteneteg) [reply molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h:262]: cool did not know!

**Review by petersteneteg — APPROVED** [review]

**Review by petersteneteg — APPROVED** [review]

**Comment by petersteneteg** [comment]
seems modules is broken at the moment... will try and fix that

**Pull request merged**